torch_geometric/nn/models/label_usage.py

```diff
@@ -0,0 +1,120 @@
+    as introduced in `"Bag of Tricks for Node Classification"
+    <https://arxiv.org/abs/2103.13355>`_ paper.
+
+    .. note::
+
+        When using the :class:`LabelUsage`, adjust the model's input dimension
+        accordingly to include both features and classes.
+
+    Args:
+        base_model: An instance of the model that will do the
+            inner forward pass.
+        num_classes (int): Number of classes in dataset
+        split_ratio (float): Proportion of true labels to use as features
+            during training (default: :obj:'0.5')
+        num_recycling_iterations (int): Number of iterations for the
+            label reuse procedure to cycle predicted soft labels
+            (default: :obj:'0')
+        return_tuple (bool): If true, returns (output, train_label_idx,
+            train_pred_idx) otherwise returns prediction output
+            (default :obj:'False')
+    """
+    def __init__(
+        self,
+        base_model: torch.nn.Module,
+        num_classes: int,
+        split_ratio: float = 0.5,
+        num_recycling_iterations: int = 0,
+        return_tuple: bool = False,
+        training: bool = True
+    ):
+
+        super().__init__()
+        self.base_model = base_model
+        self.num_classes = num_classes
+        self.split_ratio = split_ratio
+        self.num_recycling_iterations = num_recycling_iterations
+        self.return_tuple = return_tuple
+        self.training = training
+
+    def forward(
+        self,
+        feat: Tensor,
+        edge_index: Adj,
+        y: Tensor,
+        mask: Tensor,
+    ):
+        r"""Forward pass using label usage algorithm.
+
+        Args:
+            feat (torch.Tensor): Node feature tensor of dimension (N,F)
+                where N is the number of nodes and
+                F is the number of features per node
+            edge_index (torch.Tensor or SparseTensor): The edge connectivity
+                to be passed to base_model
+            y (torch.Tensor): Node ground-truth labels tensor of dimension
+                of (N,) for 1D tensor or (N,1) for 2D tensor
+            mask (torch.Tensor): A mask or index tensor denoting which nodes
+                are used during training
+        """
+        assert feat.dim() == 2, f"feat must be 2D but got shape {feat.shape}"
+        assert y.dim() == 1 or (y.dim() == 2 and y.size(1) == 1),\
+            f"Expected y to be either (N,) or (N, 1), but got shape {y.shape}"
+
+        # set unlabeled mask for unlabeled indices
+        unlabeled_mask = torch.ones(feat.size(0), dtype=torch.bool).to(feat.device)
+
+        # add labels to features for train_labels nodes if in training
+        # else fill true labels for all nodes in mask
+        # zero value nodes in train_pred
+        onehot = torch.zeros([feat.shape[0], self.num_classes]).to(feat.device)
+        if self.training:
+            # random split mask based on split ratio
+            if mask.dtype == torch.bool:
```

Comment:
Great catch on the different cases here!  I think we can make it a little shorter:
```
            if mask.dtype == torch.bool:
                mask = mask.nonzero(as_tuple=False).view(-1)  # convert to index tensor
            split_mask = torch.rand(mask.shape) < self.split_ratio
            train_labels = mask[split_mask] # D_L: nodes with labels
            train_pred = mask[~split_mask] # D_U: nodes to predict labels
```